File: clients/go/collection.go
package zydecodb

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
)

// Document is a JSON document: a map keyed by field name. The "_id" field is
// the document's string primary key.
type Document = map[string]any

// Collection is the product surface: a MongoDB-inspired collection of JSON
// documents over the binary client. Filters and updates use the familiar
// $-operators; the server plans the access path and re-checks the full filter.
type Collection struct {
	client *Client
	name   string
}

// Name returns the collection name.
func (c *Collection) Name() string { return c.name }

// CreateIndex creates a secondary index over one or more dotted field paths.
// It returns false if the index already existed.
func (c *Collection) CreateIndex(ctx context.Context, fields []string, unique bool) (bool, error) {
	parts := make([]string, len(fields))
	for i, f := range fields {
		parts[i] = strings.ReplaceAll(f, ".", "_")
	}
	name := "by_" + strings.Join(parts, "_")
	return c.client.DefineIndex(ctx, c.name, name, fields, unique, true)
}

// InsertOne inserts a document, generating "_id" if absent, and returns the id.
// It returns an error (see IsConflict) if a unique index would be violated.
func (c *Collection) InsertOne(ctx context.Context, doc Document, relaxed bool) (string, error) {
	id, _ := doc["_id"].(string)
	if id == "" {
		id = GenerateID()
	}
	cp := make(Document, len(doc)+1)
	for k, v := range doc {
		cp[k] = v
	}
	cp["_id"] = id
	body, err := json.Marshal(cp)
	if err != nil {
		return "", fmt.Errorf("zydecodb: marshal document: %w", err)
	}
	if _, err := c.client.PutDocument(ctx, c.name, id, body, relaxed); err != nil {
		return "", err
	}
	return id, nil
}

// InsertMany inserts each document in order, returning the assigned ids.
func (c *Collection) InsertMany(ctx context.Context, docs []Document) ([]string, error) {
	ids := make([]string, 0, len(docs))
	for _, d := range docs {
		id, err := c.InsertOne(ctx, d, false)
		if err != nil {
			return ids, err
		}
		ids = append(ids, id)
	}
	return ids, nil
}

// ReplaceOne inserts or fully replaces the document at docID.
func (c *Collection) ReplaceOne(ctx context.Context, docID string, doc Document, relaxed bool) (uint64, error) {
	cp := make(Document, len(doc)+1)
	for k, v := range doc {
		cp[k] = v
	}
	cp["_id"] = docID
	body, err := json.Marshal(cp)
	if err != nil {
		return 0, fmt.Errorf("zydecodb: marshal document: %w", err)
	}
	return c.client.PutDocument(ctx, c.name, docID, body, relaxed)
}

// UpdateResult summarizes an update operation.
type UpdateResult struct {
	Matched  int64 `json:"matched"`
	Modified int64 `json:"modified"`
}

func (c *Collection) update(ctx context.Context, filter, update Document, multi, relaxed bool) (UpdateResult, error) {
	var res UpdateResult
	fb, err := marshalFilter(filter)
	if err != nil {
		return res, err
	}
	ub, err := json.Marshal(update)
	if err != nil {
		return res, fmt.Errorf("zydecodb: marshal update: %w", err)
	}
	body, err := c.client.Update(ctx, c.name, fb, ub, multi, relaxed)
	if err != nil {
		return res, err
	}
	if err := json.Unmarshal(body, &res); err != nil {
		return res, fmt.Errorf("zydecodb: decode update result: %w", err)
	}
	return res, nil
}

// UpdateOne applies update to the first matching document.
func (c *Collection) UpdateOne(ctx context.Context, filter, update Document, relaxed bool) (UpdateResult, error) {
	return c.update(ctx, filter, update, false, relaxed)
}

// UpdateMany applies update to all matching documents.
func (c *Collection) UpdateMany(ctx context.Context, filter, update Document, relaxed bool) (UpdateResult, error) {
	return c.update(ctx, filter, update, true, relaxed)
}

// DeleteOne deletes the first matching document and returns the deleted count.
func (c *Collection) DeleteOne(ctx context.Context, filter Document, relaxed bool) (int64, error) {
	fb, err := marshalFilter(filter)
	if err != nil {
		return 0, err
	}
	return c.client.DeleteByFilter(ctx, c.name, fb, false, relaxed)
}

// DeleteMany deletes all matching documents and returns the deleted count.
func (c *Collection) DeleteMany(ctx context.Context, filter Document, relaxed bool) (int64, error) {
	fb, err := marshalFilter(filter)
	if err != nil {
		return 0, err
	}
	return c.client.DeleteByFilter(ctx, c.name, fb, true, relaxed)
}

// QueryOptions tunes a Find at the collection level.
type QueryOptions struct {
	Sort     []SortKey
	Include  []string // include only these fields (mutually exclusive with Exclude)
	Exclude  []string // exclude these fields
	Skip     uint32
	Limit    uint32
	PageSize uint32
}

func (q QueryOptions) projection() (Projection, error) {
	if len(q.Include) > 0 && len(q.Exclude) > 0 {
		return Projection{}, fmt.Errorf("zydecodb: projection cannot mix include and exclude fields")
	}
	if len(q.Include) > 0 {
		return Projection{Mode: ProjInclude, Fields: q.Include}, nil
	}
	if len(q.Exclude) > 0 {
		return Projection{Mode: ProjExclude, Fields: q.Exclude}, nil
	}
	return Projection{Mode: ProjNone}, nil
}

// Find returns all matching documents, decoded as Documents.
func (c *Collection) Find(ctx context.Context, filter Document, opts QueryOptions) ([]Document, error) {
	fb, err := marshalFilter(filter)
	if err != nil {
		return nil, err
	}
	proj, err := opts.projection()
	if err != nil {
		return nil, err
	}
	bodies, err := c.client.Find(ctx, c.name, fb, FindOptions{
		Sort:       opts.Sort,
		Projection: proj,
		Skip:       opts.Skip,
		Limit:      opts.Limit,
		PageSize:   opts.PageSize,
	})
	if err != nil {
		return nil, err
	}
	out := make([]Document, 0, len(bodies))
	for _, b := range bodies {
		doc := Document{}
		if len(b) > 0 {
			if err := json.Unmarshal(b, &doc); err != nil {
				return nil, fmt.Errorf("zydecodb: decode document: %w", err)
			}
		}
		out = append(out, doc)
	}
	return out, nil
}

// FindOne returns the first matching document, or (nil, nil) if none match.
func (c *Collection) FindOne(ctx context.Context, filter Document, opts QueryOptions) (Document, error) {
	opts.Limit = 1
	docs, err := c.Find(ctx, filter, opts)
	if err != nil || len(docs) == 0 {
		return nil, err
	}
	return docs[0], nil
}

// Get fetches one document directly by id (fast path), or (nil, nil) if absent.
func (c *Collection) Get(ctx context.Context, docID string) (Document, error) {
	body, err := c.client.GetDocument(ctx, c.name, docID)
	if err != nil || body == nil {
		return nil, err
	}
	doc := Document{}
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, fmt.Errorf("zydecodb: decode document: %w", err)
	}
	return doc, nil
}

// CountDocuments returns the number of documents matching filter (nil = all).
func (c *Collection) CountDocuments(ctx context.Context, filter Document) (int64, error) {
	fb, err := marshalFilter(filter)
	if err != nil {
		return 0, err
	}
	return c.client.Count(ctx, c.name, fb)
}

// Distinct returns the distinct values of field across matching documents.
func (c *Collection) Distinct(ctx context.Context, field string, filter Document) ([]any, error) {
	fb, err := marshalFilter(filter)
	if err != nil {
		return nil, err
	}
	return c.client.Distinct(ctx, c.name, field, fb)
}

// marshalFilter serializes a filter, treating a nil/empty filter as "match all"
// (empty bytes on the wire).
func marshalFilter(filter Document) ([]byte, error) {
	if len(filter) == 0 {
		return nil, nil
	}
	b, err := json.Marshal(filter)
	if err != nil {
		return nil, fmt.Errorf("zydecodb: marshal filter: %w", err)
	}
	return b, nil
}

File: clients/typescript/src/collection.ts
import type { Client, UpdateResult } from "./client.ts";
import { generateId } from "./client.ts";
import { ZydecoError } from "./errors.ts";
import { Proj, type Projection, type SortKey } from "./protocol.ts";

/** A JSON document. The "_id" field is the document's string primary key. */
export type Document = Record<string, unknown>;

export interface QueryOptions {
  sort?: SortKey[];
  /** Include only these fields (mutually exclusive with `exclude`). */
  include?: string[];
  /** Exclude these fields. */
  exclude?: string[];
  skip?: number;
  limit?: number;
  pageSize?: number;
}

const EMPTY = Buffer.alloc(0);

/**
 * The product surface: a collection of JSON documents over the
 * binary client. Filters and updates use the familiar $-operators; the server
 * plans the access path and re-checks the full filter.
 */
export class Collection {
  private readonly client: Client;
  readonly name: string;

  constructor(client: Client, name: string) {
    this.client = client;
    this.name = name;
  }

  /**
   * Create a secondary index over one or more dotted field paths. Returns false
   * if the index already existed.
   */
  createIndex(fields: string[], unique = false): Promise<boolean> {
    const indexName = "by_" + fields.map((f) => f.replaceAll(".", "_")).join("_");
    return this.client.defineIndex(this.name, indexName, fields, unique);
  }

  /** Insert a document, generating "_id" if absent. Returns the id. */
  async insertOne(document: Document, relaxed = false): Promise<string> {
    const id = typeof document._id === "string" && document._id ? document._id : generateId();
    const doc = { ...document, _id: id };
    await this.client.putDocument(this.name, id, jsonBytes(doc), relaxed);
    return id;
  }

  async insertMany(documents: Document[]): Promise<string[]> {
    const ids: string[] = [];
    for (const d of documents) ids.push(await this.insertOne(d));
    return ids;
  }

  /** Insert or fully replace the document at docId. */
  replaceOne(docId: string, document: Document, relaxed = false): Promise<bigint> {
    const doc = { ...document, _id: docId };
    return this.client.putDocument(this.name, docId, jsonBytes(doc), relaxed);
  }

  updateOne(filter: Document, update: Document, relaxed = false): Promise<UpdateResult> {
    return this.client.update(this.name, filterBytes(filter), jsonBytes(update), false, relaxed);
  }

  updateMany(filter: Document, update: Document, relaxed = false): Promise<UpdateResult> {
    return this.client.update(this.name, filterBytes(filter), jsonBytes(update), true, relaxed);
  }

  deleteOne(filter: Document, relaxed = false): Promise<number> {
    return this.client.deleteByFilter(this.name, filterBytes(filter), false, relaxed);
  }

  deleteMany(filter: Document, relaxed = false): Promise<number> {
    return this.client.deleteByFilter(this.name, filterBytes(filter), true, relaxed);
  }

  /** Return all matching documents, decoded as objects. */
  async find(filter: Document | null, opts: QueryOptions = {}): Promise<Document[]> {
    const bodies = await this.client.find(this.name, filterBytes(filter), {
      sort: opts.sort,
      projection: projection(opts),
      skip: opts.skip,
      limit: opts.limit,
      pageSize: opts.pageSize,
    });
    return bodies.map((b) => (b.length ? (JSON.parse(b.toString("utf8")) as Document) : {}));
  }

  /** Return the first matching document, or null if none match. */
  async findOne(filter: Document | null, opts: QueryOptions = {}): Promise<Document | null> {
    const docs = await this.find(filter, { ...opts, limit: 1 });
    return docs.length ? docs[0]! : null;
  }

  /** Fetch one document directly by id (fast path), or null if absent. */
  async get(docId: string): Promise<Document | null> {
    const body = await this.client.getDocument(this.name, docId);
    return body === null ? null : (JSON.parse(body.toString("utf8")) as Document);
  }

  countDocuments(filter: Document | null = null): Promise<number> {
    return this.client.count(this.name, filterBytes(filter));
  }

  distinct(field: string, filter: Document | null = null): Promise<unknown[]> {
    return this.client.distinct(this.name, field, filterBytes(filter));
  }
}

function projection(opts: QueryOptions): Projection {
  if (opts.include?.length && opts.exclude?.length) {
    throw new ZydecoError("projection cannot mix include and exclude fields");
  }
  if (opts.include?.length) return { mode: Proj.Include, fields: opts.include };
  if (opts.exclude?.length) return { mode: Proj.Exclude, fields: opts.exclude };
  return { mode: Proj.None, fields: [] };
}

function jsonBytes(value: unknown): Buffer {
  return Buffer.from(JSON.stringify(value), "utf8");
}

/** A nil/empty filter is "match all" (empty bytes on the wire). */
function filterBytes(filter: Document | null): Buffer {
  if (!filter || Object.keys(filter).length === 0) return EMPTY;
  return jsonBytes(filter);
}
